For the build issue, we need to resolve this error:
```
   Compiling pallet-elections-phragmen v2.0.0 (/Users/ericlim/project/plug-blockchain-2/frame/elections-phragmen)
   Compiling sp-consensus-babe v2.0.0 (/Users/ericlim/project/plug-blockchain-2/primitives/consensus/babe)
   Compiling pallet-contracts v2.0.0 (/Users/ericlim/project/plug-blockchain-2/frame/contracts)
error[E0407]: method `origin` is not a member of trait `Ext`
   --> /Users/ericlim/project/plug-blockchain-2/frame/contracts/src/exec.rs:763:2
    |
763 |       fn origin(&self) -> &T::AccountId {
    |  _____^
764 | |         &self.origin
765 | |     }
    | |_____^ not a member of trait `Ext`
```

--- a/frame/contracts/src/exec.rs
+++ b/frame/contracts/src/exec.rs
@@ -142,6 +142,9 @@ pub trait Ext {
 	/// Returns a reference to the account id of the caller.
 	fn caller(&self) -> &AccountIdOf<Self::T>;
 
+	/// `origin` is the account which authored the original extrinsic leading to the current contract call(s). It does not change on nested calls as with `caller`.
+	fn origin(&self) -> &AccountIdOf<Self::T>;
+
 	/// Returns a reference to the account id of the current contract.
 	fn address(&self) -> &AccountIdOf<Self::T>;
 
@@ -294,12 +297,10 @@ where
 	/// The specified `origin` address will be used as `sender` for. The `origin` must be a regular
 	/// account (not a contract).
 	pub fn top_level(origin: T::AccountId, cfg: &'a Config<T>, vm: &'a V, loader: &'a L) -> Self {
-		let source_account = origin;
-		let original_account = source_account.clone();
 		ExecutionContext {
 			parent: None,
 			self_trie_id: None,
-			self_account: source_account,
+			self_account: origin.clone(),
 			overlay: OverlayAccountDb::<T>::new(&DirectAccountDb),
 			depth: 0,
 			deferred: Vec::new(),
@@ -308,7 +309,7 @@ where
 			loader: &loader,
 			timestamp: T::Time::now(),
 			block_number: <system::Module<T>>::block_number(),
-			origin: original_account
+			origin: origin.clone()
 		}
 	}
 
@@ -968,6 +969,33 @@ mod tests {
 		});
 
 		assert_eq!(&*test_data.borrow(), &vec![0, 1]);
+	}
+
+	#[test]
+	fn origin_is_correct_at_top_level() {
+		ExtBuilder::default().build().execute_with(|| {
+			let value = Default::default();
+			let vm = MockVm::new();
+			let loader = MockLoader::empty();   
+		   	let cfg = Config::preload();
+			let mut ctx = ExecutionContext::top_level(ALICE, &cfg, &vm, &loader);	 
+			
+			assert_eq!(*ctx.new_call_context(BOB, value).origin(), ALICE);
+		});
+	}
+	
+	#[test]
+	fn origin_is_correct_for_nested_calls() {
+		ExtBuilder::default().build().execute_with(|| {
+			let value = Default::default();
+			let vm = MockVm::new();
+			let loader = MockLoader::empty();   
+		   	let cfg = Config::preload();
+			let ctx = ExecutionContext::top_level(ALICE, &cfg, &vm, &loader);	 
+			let mut nested = ctx.nested(BOB, None);
+
+			assert_eq!(*nested.new_call_context(CHARLIE, value).origin(), ALICE);
+		});
 	}
 
 	#[test]

--- a/frame/contracts/src/exec.rs
+++ b/frame/contracts/src/exec.rs
@@ -991,10 +991,10 @@ mod tests {
 			let vm = MockVm::new();
 			let loader = MockLoader::empty();   
 		   	let cfg = Config::preload();
-			let ctx = ExecutionContext::top_level(ALICE, &cfg, &vm, &loader);	 
-			let mut nested = ctx.nested(BOB, None);
+			let ctx = ExecutionContext::top_level(BOB, &cfg, &vm, &loader);	 
+			let mut nested = ctx.nested(ALICE, None);
 
-			assert_eq!(*nested.new_call_context(CHARLIE, value).origin(), ALICE);
+			assert_eq!(*nested.new_call_context(CHARLIE, value).origin(), BOB);
 		});
 	}
 

--- a/frame/contracts/src/wasm/mod.rs
+++ b/frame/contracts/src/wasm/mod.rs
@@ -280,6 +280,9 @@ mod tests {
 		fn caller(&self) -> &u64 {
 			&42
 		}
+		fn origin(&self) -> &u64 {
+			&24
+		}
 		fn address(&self) -> &u64 {
 			&69
 		}
@@ -380,6 +383,9 @@ mod tests {
 		}
 		fn caller(&self) -> &u64 {
 			(**self).caller()
+		}
+		fn origin(&self) -> &u64 {
+			(**self).origin()
 		}
 		fn address(&self) -> &u64 {
 			(**self).address()

--- a/frame/contracts/src/wasm/mod.rs
+++ b/frame/contracts/src/wasm/mod.rs
@@ -779,6 +779,68 @@ mod tests {
 		).unwrap();
 	}
 
+	/// calls `ext_origin`, loads the address from the scratch buffer and
+	/// compares it with the constant 24.
+	const CODE_ORIGIN: &str = r#"
+(module
+	(import "env" "ext_origin" (func $ext_origin))
+	(import "env" "ext_scratch_size" (func $ext_scratch_size (result i32)))
+	(import "env" "ext_scratch_read" (func $ext_scratch_read (param i32 i32 i32)))
+	(import "env" "memory" (memory 1 1))
+
+	(func $assert (param i32)
+		(block $ok
+			(br_if $ok
+				(get_local 0)
+			)
+			(unreachable)
+		)
+	)
+
+	(func (export "call")
+		;; fill the scratch buffer with the caller.
+		(call $ext_origin)
+
+		;; assert $ext_scratch_size == 8
+		(call $assert
+			(i32.eq
+				(call $ext_scratch_size)
+				(i32.const 8)
+			)
+		)
+
+		;; copy contents of the scratch buffer into the contract's memory.
+		(call $ext_scratch_read
+			(i32.const 8)		;; Pointer in memory to the place where to copy.
+			(i32.const 0)		;; Offset from the start of the scratch buffer.
+			(i32.const 8)		;; Count of bytes to copy.
+		)
+
+		;; assert that contents of the buffer is equal to the i64 value of 24.
+		(call $assert
+			(i64.eq
+				(i64.load
+					(i32.const 8)
+				)
+				(i64.const 24)
+			)
+		)
+	)
+
+	(func (export "deploy"))
+)
+"#;
+
+	#[test]
+	fn origin() {
+		let _ = execute(
+			CODE_ORIGIN,
+			vec![],
+			MockExt::default(),
+			&mut GasMeter::with_limit(50_000, 1),
+		).unwrap();
+	}
+
 	/// calls `ext_address`, loads the address from the scratch buffer and
 	/// compares it with the constant 69.
 	const CODE_ADDRESS: &str = r#"

--- a/frame/contracts/src/wasm/runtime.rs
+++ b/frame/contracts/src/wasm/runtime.rs
@@ -546,6 +546,13 @@ define_env!(Env, <E: Ext>,
 		Ok(())
 	},
 
+	// Stores the the origin address of the extrinsic into the scratch buffer.
+	ext_origin(ctx) => {
+		ctx.scratch_buf.clear();
+		ctx.ext.origin().encode_to(&mut ctx.scratch_buf);
+		Ok(())
+	},
+
 	// Stores the address of the current contract into the scratch buffer.
 	ext_address(ctx) => {
 		ctx.scratch_buf.clear();

